src/java/org/apache/cassandra/db/commitlog/AbstractCommitLogSegmentManager.java

```diff
@@ -166,7 +148,22 @@ public void runMayThrow() throws Exception
             }
         };
 
+        BufferType BBType = SimpleCachedBufferPool.DEFAULT_PREFERRED_BB_TYPE;
+        if (commitLog.configuration.useEncryption())
+            // Keep reusable buffers on-heap regardless of compression preference so we avoid copy off/on repeatedly during decryption
+            // Also: we want to keep the compression buffers on-heap as we need those bytes for encryption,
+            // and we want to avoid copying from off-heap (compression buffer) to on-heap encryption APIs
+            BBType = BufferType.ON_HEAP;
+        else if (commitLog.configuration.useCompression())
+            BBType = commitLog.configuration.getCompressor().preferredBufferType();
+
+        synchronized(this)
+        {
+            this.bufferPool = new SimpleCachedBufferPool(DatabaseDescriptor.getCommitLogMaxCompressionBuffersInPool(), DatabaseDescriptor.getCommitLogSegmentSize(), BBType);
```

Comment:
We can make this more 'solid' by closing the pool before re-creation to defend against `start()` calls without a matching ' close()`.